database/pebble/db.go

```diff
@@ -0,0 +1,463 @@
+	}
+	ret := slices.Clone(data)
+	return ret, closer.Close()
+}
+
+func (db *Database) Put(key []byte, value []byte) error {
+	// Put causes panic if the db has already been closed
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	// Use of [pebble.NoSync] here means we don't wait for the [Set] to be
+	// persisted to the WAL before returning. Basic benchmarking indicates that
+	// waiting for the WAL to sync reduces performance by 20%.
+	return updateError(db.db.Set(key, value, pebble.NoSync))
+}
+
+func (db *Database) Delete(key []byte) error {
+	// Delete causes panic if the db has already been closed
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	return updateError(db.db.Delete(key, pebble.NoSync))
+}
+
+func (db *Database) Compact(start []byte, limit []byte) error {
+	// Pebble Compact causes panic if the db has already been closed.
+	// Avalanche database just return error instead.
+	if db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	if bytes.Equal(start, limit) && limit != nil {
+		// The default compare function of pebble is bytes.Compare.
+		// Use this default compare function since we don't setup
+		// the compare function when we create pebble db.
+		// It's impossible to retrieve this default
+		// compare function from the pebble package since
+		// it's defined in internal.
+		// We use bytes.Equal instead of bytes.Compare
+		// because golangci_lint recommends
+
+		// pebble return error if start and limit are the same,
+		// even if both are nil.
+		// Do nothing just ignore the error if start and limt are the same
+		// and not nil since there is no need to compact and avalanche db
+		// expects a nil return.
+		return nil
+	} else if limit != nil {
+		return updateError(db.db.Compact(start, limit, true))
+	}
+
+	// A nil limit is treated as a key after all keys in avalanche DB.
+	// But pebble treats a nil, no matter start or limit, as a key before
+	// all keys in the DB
+	it := db.db.NewIter(&pebble.IterOptions{})
+	if it.Last() {
+		if lastkey := it.Key(); lastkey != nil {
+			return updateError(db.db.Compact(start, lastkey, true))
+		}
+	}
+
+	return database.ErrNotFoundLastKey
+}
+
+// batch is a wrapper around a pebbleDB batch to contain sizes.
+type batch struct {
+	batch *pebble.Batch
+	db    *Database
+	size  int
+
+	// Support batch rewrite
+	applied atomic.Bool
+}
+
+func (db *Database) NewBatch() database.Batch {
+	return &batch{
+		db:    db,
+		batch: db.db.NewBatch(),
+	}
+}
+
+func (b *batch) Put(key, value []byte) error {
+	b.size += len(key) + len(value) + pebbleByteOverHead
+	return b.batch.Set(key, value, pebble.NoSync)
+}
+
+func (b *batch) Delete(key []byte) error {
+	b.size += len(key) + pebbleByteOverHead
+	return b.batch.Delete(key, pebble.NoSync)
+}
+
+func (b *batch) Size() int { return b.size }
+
+func (b *batch) Write() error {
+	// write causes panic if the db has already been closed
+	if b.db.closed.Get() {
+		return database.ErrClosed
+	}
+
+	// Support batch rewrite
```

Comment:
Maybe a dumb question but what's the use-case for a batch rewrite?

Reply:
🤷 The database.Database tests enforce this